let error = table
error.msg = "error"

let arg_mismatch = table from error
arg_mismatch.msg = "argument mismatch"

let uncallable = table from error
uncallable.msg = "uncallable value"

let interrupt = table from error
interrupt.msg = "received interrupt signal"

let fpe = table from error
fpe.msg = "floating point error"

let segfault = table from error
segfault.msg = "segmentation fault"

export error as foreign "rain_exc_error"
export arg_mismatch as foreign "rain_exc_arg_mismatch"
export fpe as foreign "rain_exc_fpe"
export interrupt as foreign "rain_exc_interrupt"
export segfault as foreign "rain_exc_segfault"
export uncallable as foreign "rain_exc_uncallable"

let safe = func(val)
  while val is table
    break if !(val.metatable)
    val = val.metatable

  return val != error
